# Short Read Taxonomic Classification

Week 3 goals are to understand the tools and steps involved in taxonomic classification of metagenomic short reads.

![Workflow](metagenomics_flow_classify.png)


## kraken2 Manuscript

[kraken2](https://pubmed.ncbi.nlm.nih.gov/31779668/) is tool that uses a k-mer based approach to assign a taxonomic classification to sequencing reads. Classification estimates can be refined with the bayesian tool [bracken](https://peerj.com/articles/cs-104/). Downloading and installing kraken2 databases are the most time-intensive parts of this step. From there, 


## Classify Reads

Most of the following steps are abstracted from the [kraken2 manual](https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown), which should be considered the primary source for questions/troubleshooting.

### Download kraken2 database

The kraken2 database gets updated every few months to reflect new taxa, new reference sequences, and updated taxonomic nomenclature (sigh). At time of writing this GitHub page, the pre-built databases were moved to an amazon server with this [public-facing site](https://benlangmead.github.io/aws-indexes/k2). 

On this site, note there are a number of different pre-built databases that you can download and setup depending on your data and research questions. For our purposes, to try to minimize our disk usage on this server and to demonstrate the steps, we will work with the standard-8 database, which is a version that is limited to <8gb. If you want to convince yourself of whether the taxa in this limited database will be useful, you can view the results of the [database inspection output](https://genome-idx.s3.amazonaws.com/kraken/standard_08gb_20221209/inspect.txt).

The kraken2 manual includes details for how to download publicly-available data to build your own database or add custom sequences of interest to your database but these are beyond our scope.

``` console
# download most recent 8gb database at time of GitHub commit
wget https://genome-idx.s3.amazonaws.com/kraken/k2_standard_08gb_20221209.tar.gz
```
``` console
# decompress database
tar -xf k2_standard_08gb_20221209.tar.gz
```

### Run kraken2

Once we have downloaded a pre-built database, we can initialize our variables and run kraken2.

*Note: There are multiple output options for kraken2 but bracken requires the default report format from kraken2 (which we have labeled with the .kreport suffix in our command below).*

``` console
indir=/home/mwoodwo/prj/PractMeta/01.trimmomatic
outdir=/home/mwoodwo/prj/PractMeta/02.kraken2
ID=PM6-S
db=/home/mwoodwo/software/kraken2-db/8gb_230206
```

``` console
kraken2 --db $db --threads 20 --output ${ID}.kraken2 --report ${ID}.kreport --paired ${indir}/${ID}_P1.fastq ${indir}/${ID}_P2.fastq
```

We can also loop over a list of IDs to run kraken2 for several files.
``` console
IDlist=
for ID in `cat $IDlist`; do echo starting $ID ...; kraken2 --db $db --threads 20 --output ${ID}.kraken2 --report ${ID}.kreport --paired ${indir}/${ID}_P1.fastq ${indir}/${ID}_P2.fastq; echo $ID complete ...; echo ''; done
```

### Set up bracken

Most of the following steps are abstracted from the [bracken](https://github.com/jenniferlu717/Bracken) manual, which should be considered the primary source for troubleshooting.



### Summarize output

- combine output files
- kraken-biom

---

## Week 4 prep (5 minutes)

Next week, we will work on diversity metrics with taxonomy matrices. To prepare for next week, let's:

- Download PREMIX kraken2/bracken output files
- Review [Vegan](https://github.com/vegandevs/vegan) R package
- Review [Phyloseq](https://joey711.github.io/phyloseq/index.html) R package
