# Coverage visualization

Week 9 goals are to understand the tools and steps involved in creating plots of metagenomic coverage analyses.

## 1. Genes & Genomes

We are going to be creating plots in R with ggplot2 and tidyverse. We will highlight some techniques to simplify creating multiple plots at once for exploratory analysis and communicating your results.

The data we will focus on are from two separate runs of inStrain.

In the first run, we calculated genome coverage statistics for three bacterial genomes that were sequenced from isolates cultured from baseline stool samples from PREMIX participant PM06.

In the second run, we calculated gene coverage statistics for contigs to get *in situ* coverage statistics for genes of interest (namely antimicrobial resistance genes in our case).

## 2.  Summarize inStrain


### Metagenome results
Sample output from the second run are uploaded to our OneDrive folder under Data/PM06_inStrain_metagenome_results.

To create a summary matrix of coverage depth for each gene normalized by gene length and number of genome equivalents per metagenome (RPKG, reads-per-kilobase-per-genome-equivalent), we can use a custom python script to parse the output from several tools (inStrain gene_info.tsv files, filtered AMRFinder results, and genome equivalent estimates from MicrobeCensus).

```console
######## define variables
gene_info_files=  #typically you would need to link or copy the output from each individual 
                  #inStrain profile/output/*gene_info.tsv file into a new directory
AMRFinder_tsvs=   #this script expects a directory containing one filtered AMRFinder output 
                  #table per metagenome as described in earlier weeks
microbe_census_summary= #this script expects a summary tsv file of all microbe census results for each metagenome
outdir=           #where to write output

script_path=      #where you have saved the script 02_amrfinder_validate_and_summarize_RPKG.py
                  #if you want to store this as a variable
```

```console
python $script -a $AMRFinder_tsvs -i $gene_info_files -m $microbe_census_summary -o $outdir -s -V -g -r -v

# you can explore the options of this script with python $script -h
# although for now, you have to always use the -r option (you get what you pay for)
```

Now change into your output directory to examine the output.


### Genome results
Sample output from the second run are uploaded to our OneDrive folder under Data/PM06_inStrain_genome_results.

To create a summary matrix of coverage depth for each gene normalized by gene length and number of genome equivalents per metagenome (RPKG, reads-per-kilobase-per-genome-equivalent), we can use a custom python script to parse the output from several tools (inStrain gene_info.tsv files, filtered AMRFinder results, and genome equivalent estimates from MicrobeCensus).

```console
######## define variables
genome_info_files=  #typically you would need to link or copy the output from each individual 
                  #inStrain profile/output/*genome_info.tsv file into a new directory
outfile=           #path and name for tsv summary file to write (e.g. ${outdir}/PM06_summary.tsv)

script_path=      #where you have saved the script summarize_inStrain_genome_info_files.py
                  #if you want to store this as a variable
```

```console
python $script -i $genome_info_files -o $outfile -v

# you can explore the options of this script with python $script -h
```

Now change into your output directory to examine the output.

## 3. Creating manual color palletes in ggplot2

```r

############# specify default color scales
scale_colour_discrete <-  function(...){
  scale_colour_viridis(discrete=TRUE, ...)
}
scale_fill_discrete <-  function(...){
  scale_fill_viridis(discrete=TRUE, ...)
}


############## set default theme

theme_set(theme_classic())

############# create manual color palettes

isolate_colors <- scale_color_manual(
                      breaks = c("ESBL",
                                 "CRE",
                                 "VRE"),
                      values = c("#44546A",
                                 "orange",
                                 "darkgrey"),
                      labels = c("ESBL *K. pneumoniae*",
                                 "CR *K. pneumoniae*",
                                 "VR *E. faecium*")
)


# prep column annotations for pheatmap
PM06md <- md %>% 
            filter(Sample %in% visit_order)


c_annotation <- data.frame(PM06md$Exposure,
                           as.factor(PM06md$`MDRO Status`))
    rownames(c_annotation) <- PM06md$Sample
    colnames(c_annotation) <- c("Exposure",
                                "MDRO Status")


# define annotation colors
meta_colors <- list(
          Exposure=c(Donor="#B6854D",
                 FMT="#79402E",
                 None="#D9D0D3",
                 Prep="#9986A5"),
          # Group=c(Donor="#000004FF",
          #         FMT="#CC4678FF",
          #         Screen="#F0F921FF"),
          `MDRO Status`=c(Negative="#46ACC8",
                  Positive="#E58601")
          )    
    
    
```

## 4. Creating plotting functions in ggplot2

```r
# define depth plotter for inStrain genome coverage summary data
depth_plotter <- function(genome_coverage_df, sample_ID){

                    depth <- genome_coverage_df %>% 
                                filter(str_detect(MG_ID, sample_ID)) %>% 
                                         ggplot(aes(x=MG_ID,
                                                    y=depth, 
                                         group=g, color=g)) + 
                              geom_line(size=2,
                                        alpha=0.8) +
                              geom_point(alpha=0.8) +
                              isolate_colors +
                              labs(
                                    x=paste(sample_ID, "Stool Metagenome"),
                                    y="Coverage Depth"
                                   ) +
                              theme(axis.text.x = element_text(angle=45,
                                                               hjust=1),
                                    legend.title = element_blank(),
                                    legend.position = "bottom",
                                    legend.text = element_markdown()) +
                              theme(axis.title.x = element_blank())
                  return(depth)
}


# define depth plotter for inStrain genome coverage summary data
breadth_plotter <- function(genome_coverage_df, sample_ID){

  breadth <- genome_coverage_df %>% 
            filter(str_detect(MG_ID, sample_ID)) %>% 
                     ggplot(aes(x=MG_ID,
                                y=breadth *100,
                     group=g, color=g)) + 
          geom_line(size=2, alpha = 0.8)  +
          geom_point(alpha = 0.8) +
          isolate_colors +
          ylim(0,100) +
          labs(y="Coverage Breadth (%)",
               x=paste(sample_ID, "Stool Metagenome")) + 
                theme(axis.text.x = element_text(angle=45,
                                           hjust=1)) +
                theme(axis.text.x = element_markdown(),
                      legend.position = "bottom",
                      legend.text = element_markdown(),
                      legend.title = element_blank()) +
          geom_hline(yintercept = 50, 
                     linetype="dashed", 
                     color= "red")


  return(breadth)
}

```


## 5. Looping plotting functions over a list

```r

library(glue)

IDs <- c("PM06", "SD01")

for(i in IDs){
  print(glue("plotting {i}"))
  print(breadth_plotter(gs,i))
  print(depth_plotter(gs,i))
}

```

## 6. Plotting coverage for PM06


## 7. Interpret results

- How did coverage change for the isolate genomes over time?
- How did normalized AMR gene coverage (RPKG) change after FMT?
- Did any AMR genes *increase* after FMT? 
- Was there any coverage of isolate genomes by the FMT dose metagenome?

## Week 10 prep (5 minutes)

Next week, we will shift gears to focus on 16S rRNA gene analyses in QIIME2.
