# Contig annotation

Week 7 goals are to understand the tools and steps involved in predicting and annotating genes on metagenome contigs.

## 1. Gene Prediction and Annotation

Gene prediction describes the bioinformatic work in detecting nucleotide sequences that are most likely to be translated into proteins with the use of multiple heuristics to identify start/stop loci etc. Prodigal is a computationally efficient and fast tool that is one of the most widely used for this step. Prodigal is able to take a fasta file containing contigs, and provide fasta output with gene-coding nucleotide and translated amino acid sequences. In addition, the user can optionally output gff files, which include additional data about the position of these genes.

The user can then use these gene fasta files as input for annotation tools that may use alignment search and/or hidden markov model profile search strategies to classify the identity for a specific gene if it is known. A recognized limitation of this approach is that even in highly studied organisms, such as *E. coli*, that many genes that are predicted may not have a known or experimentally confirmed function. These predicted genes without annotation are often classified as hypothetical genes. We will focus on annotation of antimicrobial resistance genes as an example that is relevant to many of our areas of interest and list some additional tools to explore for functional annotation.

- [prodigal paper](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-119)
- [AMRFinder plus paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8208984/)

## 2. Running prodigal

Prodigal runs pretty quickly and should be feasible for us to run prodigal as a loop for all PMO6 and donor metagenome contigs.

Let's first create a symbolic link of these contigs in a new directory within our metaSPAdes output directory for simplicity.

``` console
indir=        # directory containing all metaSPAdes output directories (e.g. 03.metaSPAdes)
outdir=       # directory in which to create new symbolic links, you will need to create this (e.g. 03.metaSPAdes/contig_links) 
IDlist=       # list of IDs over which to loop

cd $outdir

for ID in `cat $IDlist`; do echo starting $ID ...; ln -s ${indir}/${ID}/contigs.fasta ./${ID}_contigs.fasta; echo $ID complete ...; echo ''; done

```

``` console
######### define variables
indir=                        # directory containing metaSPAdes contig links
outdir=                       # directory to write prodigal output, I often use 03.prodigal
ID=                           # pick one from the list of PM06 IDs when running an individual file
args='-p meta'                # for metagenome gene prediction

```

``` console
# sanity check
echo 'ID=' $ID
echo 'indir=' $indir
echo 'outdir=' $outdir

```

``` console
# run prodigal for one ID
prodigal -a ${outdir}/${ID}.faa -d ${outdir}/${ID}.fna -f gff -i ${contig} -o ${outdir}/${ID}.gff $args
```

``` console
# run prodigal as a loop over all IDs
for ID in `cat $IDlist`; do echo starting $ID ...; contig=${indir}/${ID}_contigs.fasta; prodigal -a ${outdir}/${ID}.faa -d ${outdir}/${ID}.fna -f gff -i ${contig} -o ${outdir}/${ID}.gff $args; echo $ID complete ...; echo '---------------------'; echo ''; done
```

## 3. Annotation

We are using annotation of antimicrobial resistance (AMR) genes as a focused example of gene annotation. AMRFinder is a robust tool that combines multiple search strategies to annotate genes using a well-maintained reference database of AMR, virulence, and stress genes. It can be installed with conda followed by download and set up of the latest reference database. I recommend following the [installation instructions on their site](https://github.com/ncbi/amr/wiki/Installing-AMRFinder).


``` console
######### define variables
indir=                        # directory containing prodigal output
outdir=                       # directory to write AMRFinder output, I often use 04.AMRFinder
ID=                           # pick one from the list of PM06 IDs when running an individual file
protein=${indir}/${ID}.faa    # amino acid sequence fasta file from prodigal
```

``` console
# run AMRFinder
amrfinder -p ${protein} --plus -o ${outdir}/${ID}_amrfinder.tsv --log ${outdir}/logs/${ID}.log
```

### OPTIONAL

After running AMRFinder, you may find it useful to filter your results, for example to just focus on AMR genes and not stress genes, or only to include complete genes and exclude partial genes predicted by prodigal. You may also wish to create a binary matrix that shows which genes were detected or not within a sample that can be compared as a summary plot. You can use the following scripts for these filter and summary steps.

First, download these python scripts and store them in a directory that is in your path:

``` console
# filter 
[https://github.com/michaelwoodworth/AMRFinder_scripts/blob/master/00_amrfinder_filter.py](https://github.com/michaelwoodworth/AMRFinder_scripts/blob/master/00_amrfinder_filter.py)

# create binary matrix
[https://github.com/michaelwoodworth/AMRFinder_scripts/blob/master/01_amrfinder_binary_matrix.py](https://github.com/michaelwoodworth/AMRFinder_scripts/blob/master/01_amrfinder_binary_matrix.py)
```

Then define variables for your context:
``` console
######### define variables
indir=                        # directory containing AMRFinder output
outdir=                       # directory to write filtered tsv and summary matrix output
ID=                           # pick one from the list of PM06 IDs when running an individual file
protein=${indir}/${ID}.faa    # amino acid sequence fasta file from prodigal
```

And run them:

**1. 00_amrfinder_filter.py** - this python script was written because AMRFinder produces some hits that are incomplete genes that may reduce confidence of your results.  This may be fine in an exploratory analysis, but you may prefer to filter only hits that AMRFinder classifies as "ALLELE", "EXACT", "BLASTX", "HMM", which is the default usage for this filter script.

- *add_partial_end*: 
This script allows users to also include AMRFinder hits that were partial but located at the end of a contig sequence, which could be consistent with a sequencing/assembly issue of a gene that may be complete in host.  This option is flagged with the -m add_partial_end option.

- *just_amr*:
The -j/--just_AMR flag filters and writes a tsv file with just AMR results.

- *virulence_stress*:
The -v/--virulence_stress flag filters and writes a tsv file with non-AMR results.

```console
usage: 00_amrfinder_filter.py [-h] -i  -o  [-m] [-j] [-v]

Filter AMRFinder Plus results for high confidence matches.

This script filters AMRFinder output tables for matches, with
default criteria focused on high quality & complete matches.
e.g. >90% identity, >90% match length.

Script options also allow filtering for just AMR determinants,
or conversely, only non-AMR results (e.g. virulence/stress).

optional arguments:
  -h, --help            show this help message and exit
  -i , --input          Please specify AMRFinder input tsv file name & path.
  -o , --output         Please specify AMRFinder filtered prefix & path for
                        output tsv.
  -m , --method         Please specify filtered AMRFinder output tsv file name
                        & path. Select from: complete -or- add_partial_end
  -j, --just_AMR        Flag to write tsv with just AMR results
  -v, --virulence_stress
                        Flag to write tsv without AMR results (e.g. filter
                        only virulence, stress)

```

```console
# Example usage
for ID in `cat ${list_pth}/IDlist.txt`; do echo $ID; python ${script_path}/00_amrfinder_filter.py -i ${indir}/${ID}_amrfinder.tsv -o ${outdir}/${ID} -m add_partial_end -j -v; done
```

**01_amrfinder_binary_matrix.py** - this python script searches for .tsv files in an input directory and produces a binary presence/absence matrix for all genes across all samples coded as 0 for absent and 1 as present.

This step is useful for genomes or metagenomes if relative abundance data aren't of primary interest/available.

*Note: using relative paths for writing the output tsv can throw an error.  To avoid this, use full absolute paths or environmental variables instead.*

```console
usage: 01_amrfinder_binary_matrix.py [-h] -i INPUT -o OUTPUT [-v]

Create summary matrix of AMRFinder Plus results for plots & analysis.

This script sumamrizes filtered tables from 00_amrfinder_filter.

optional arguments:
  -h, --help      show this help message and exit
  -i , --input    Please specify input directory path.
  -o , --output   Please specify output filename & path.
  -v, --verbose   Increase output messaging detail, print results.
```

```console
# Example usage
python ${script_path}/01_amrfinder_binary_matrix.py -i ${00_filtered_path} -o ${01_output_path}/01_binary_matrix.tsv
```

## Week 8 prep (5 minutes)

Next week, we can either calculate coverage statistics with inStrain or pivot to data visualization with ggplot2. We'll discuss when we meet.
