# Contig Assembly and binning

Week 6 goals are to understand the tools and steps involved in metagenome contig assembly and binning.

## 1. metaSPAdes & binning

Assembly describes the computational work of constructing nucleic acid sequences that are as long and accurate as possible. This is one of the most computationally intensive steps in metagenomic analysis. We will read the metaSPAdes paper (and try to figure out which letters are capitalized) for background and try to run some assemblies on the server. If we have major challenges, we can work with some pre-assembled contigs from these samples.

- [metaSPAdes paper](https://doi.org/10.1101/gr.213959.116)

After assembly, we can use the contigs of a certain minimum length (usually >1,000 bp but definitely some margin longer than our short  reads) to try to group assembled contigs. You can get a sense of the quality of your metagenome assembly with [MetaQUAST](https://doi.org/10.1093/bioinformatics/btv697), which will provide some statistics like length and can compare the output of multiple assemblers if you'd like (but this is beyond our scope). Commonly used binning tools like metabat2 and maxbin2 then use statistics like coverage depth, tetranucleotide frequency, GC content, etc. to then group similar contigs into 'bins'. Bins can then be further evaluated with [CheckM](https://ecogenomics.github.io/CheckM/) using [quality statistics](https://doi.org/10.1038/nbt.3893) like completeness (what proportion of single copy genes are present?) and contamination (how many extra or unexpected single copy genes are present?). If you are feeling good about these results, you may want to then assign your bin a taxonomic classification, which can be done with tools like [GTDBTK](https://gtdb.ecogenomic.org) or [MiGA](http://microbial-genomes.org). By now, you have created and characterized a metagenome-assembled genome (MAG).

- [Read Meren's blog on the fate of contigs](https://merenlab.org/2020/01/02/visualizing-metagenomic-bins/)
- [Read the DAS Tool paper](https://doi.org/10.1038/s41564-018-0171-1)
- [Read the MetaBAT 2 paper](https://doi.org/10.7717%2Fpeerj.7359)
- [Read the MaxBin2 paper](https://doi.org/10.1093/bioinformatics/btv638)

## 2. Running metaSPAdes

We may want to just plan on starting with assembly of one PM06 file since this may not work well for us as a group to assemble all week.

First, install and activate a new conda environment with SPAdes (which comes with metaSPAdes).

``` console
######### define variables
indir=        # directory containing pre-processed reads, .gz compressed should be ok
outdir=       # directory to write metaSPAdes output, I often use 02.metaSPADes
ID=           # pick one from the list of PM06 IDs

# paired & unpaired, host decontaminated reads
R1=${indir}/${ID}_P1.fastq
R2=${indir}/${ID}_P2.fastq
U=${indir}/${ID}_U.fastq      # SPAdes will try to internally account for all reads and 
                              # may produce an error/warning if not provided the unpaired
                              # reads from trimmomatic as a concanated file

args="--meta -t 20"
```

``` console
# sanity check
echo 'ID=' $ID
echo 'indir=' $indir
echo 'outdir=' $outdir

echo '-----------------------------------'
echo 'Unpaired reads (U)=' $U
echo 'Paired forward reads (R1)=' $R1
echo 'Paired reverse reads (R2)=' $R2

```

``` console
# run metaSPAdes
spades.py $args -1 $R1 -2 $R2 -s $U -o ${outdir}/${ID}
```

## 3. Binning

- Run MaxBin2
- Run MetaBat2
- [BONUS] Automatically compare / refine bins from multiple tools with DAS Tool
- [BONUS] Assess bin quality with CheckM
- [BONUS] Classify bins with GTDBTK

## Week 7 prep (5 minutes)

Next week, we will work on contig annotation and analysis. To prepare for next week, let's:

- Read the [AMRFinder paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8208984/)
