# Introduction to Dimensionality Reduction

Week 5 goals are to understand the tools and steps involved in presenting and clustering high-dimensional data.

You will want to try to create these files on your own, but for efficiency I have uploaded files from the expected output from last week:

- [PM06_combined_Bracken_G.tsv](PM06_combined_Bracken_G.tsv): Combined output of genus level Bracken results
- [PM06_combined_Bracken_S.tsv](PM06_combined_Bracken_S.tsv): Combined output of species level Bracken results
- [PM06_mg_S-biom.json](PM06_mg_S-biom.json): Biom formatted output of species level Bracken results
- [Sample data / Metadata](PM06_metadata.tsv): Data for PM06 metagenomes and FMT dose

## 1. phyloseq package documentation

- [phyloseq full stack](https://f1000research.com/articles/5-1492/v1)

## 2. Dimensionality reduction

The goal of dimensionality reduction is a way to summarize high dimensional data (i.e. many rows, many columns) in a way that is meaningful and statistically appropriate.

## 3. Bray-Curtis PCoA Analysis in Phyloseq

We will perform these steps in phyloseq since they make it fairly simple to do so. Since phyloseq is mostly a way to consistently structure the typical components of microbiome datasets and includes many additional R packages, you could (and probably should at some point!) do this with each component step but that is outside of our scope.

``` r
# load phyloseq
library(phyloseq)

# 

```

## Week 6 prep (5 minutes)

Next week, we will work on contig assembly and binning. To prepare for next week, let's:

- Read the [metaSPAdes paper](https://pubmed.ncbi.nlm.nih.gov/28298430/) 
