# Coverage calculations

Week 8 goals are to understand the tools and steps involved in calculating coverage statistics for genomes and genes with metagenomes of interest.

## 1. Calculating coverage statistics

Coverage describes how well represented a sequence (i.e. gene or genome) is within a sequencing file.

- [inStrain paper](https://doi.org/10.1038/s41587-020-00797-0)
- [MicrobeCensus paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0611-7)

## 2. Running inStrain

To run inStrain, you need a metagenome, reference [e.g. genome, set of genomes, or large set of genomes (i.e. a database)], and typically also a file containing genes predicted from your reference. If you start with the genome(s) and metagenome of interest, the [detailed inStrain documentation](https://instrain.readthedocs.io/en/latest/index.html) can help you run inStrain to calculate coverage statistics.


## 3. Running Microbe Census

Microbe Census is a tool that can estimate the average genome size and number of genome equivalents in your metagenome. This is useful as a normalization factor for coverage estimates because the size of the genomes in a metagenome could influence how well you have accurately sampled your population of interest.

``` console
######### define variables
indir=                        # directory containing paired reads from trimmomatic
outdir=                       # directory to write MicrobeCensus output, e.g. 06.microbecensus
ID=                           # pick one from the list of PM06 IDs when running an individual file
args="-n 100000000 -t 16"     # sample 100,000,000 million reads and run with 16 threads

######### at this point, you may have compressed your reads or changed your 
######### labeling if your read naming matches one of the conventions below, 
######### this will assign it to the R1 and R2 variable.

      # set reads by indir/id
        if [[ -f "${indir}/${ID}_1.fastq.gz" ]]; then
                
                R1=${indir}/${ID}_1.fastq.gz
                R2=${indir}/${ID}_2.fastq.gz

        elif [[ -f "${indir}/${ID}_P1.fastq.gz" ]]; then

                R1=${indir}/${ID}_P1.fastq.gz
                R2=${indir}/${ID}_P2.fastq.gz

        elif [[ -f "${indir}/${ID}_paired_1.fastq" ]]; then

                R1=${indir}/${ID}_paired_1.fastq
                R2=${indir}/${ID}_paired_2.fastq

        else
                R1=${indir}/${ID}_P1.fastq
                R2=${indir}/${ID}_P2.fastq

        fi

```

``` console
run_microbe_census.py $args ${R1},${R2} ${outdir}/${ID}.census
```

The resulting MicrobeCensus files can then be summarized with the [summarize_microbecensus.py](../scripts/summarize_microbecensus.py) script into a single tsv.

## 4. Summarize genome and normalized gene coverage

After running inStrain profile, you will find an output directory in each .IS folder containing genome_info.tsv and gene_info.tsv files, which contain raw coverage data. We can parse these files with python scripts, link them with our (filtered) AMRFinder gene annotations and MicrobeCensus estimates of genome equiavelents to calculate RPKG (reads per kilobase [gene length] per genome equivalents in a metagenome for AMR genes of interest.

*Note: these python scripts require python >3.6, as well as the numpy and pandas libraries.*

I typically will link these output files in a new directory so they are all in one place with something like this (you will need to adjust the filenames to match your context):

``` console
for ID in `cat $IDlist`; do ln -s ../${ID}_${ID}.IS/output/${ID}_reference.IS_gene_info.tsv; echo $ID gene_info.tsv linked; done
for ID in `cat $IDlist`; do ln -s ../${ID}_${ID}.IS/output/${ID}_reference.IS_genome_info.tsv; echo $ID genome_info.tsv linked; done
```

### 4a. Summarize genome coverage

``` console
######### define variables
script="${pathtoscripts}/summarize_inStrain_genome_info_files.py"
ID=
indir=      # path to linked genome_info.tsv files
outfile=    # path to write coverage summary file e.g. ${outdir}/PM06_genome_info_summary.tsv

```

``` console
$script -i $indir -o $outfile -v

```

### 4b. Summarize gene coverage

- (Optional) Filter AMRFinder tables for gene subsets (e.g. complete, AMR genes only)

- Run RPKG script
``` console
######### define variables
amrfinder_tsv_path= 
instrain_geneinfo_tsv_path = 
MicrobeCensus_summary_file =
outdir=
```

``` console
02_amrfinder_validate_and_summarize_RPKG.py -a $amrfinder_tsv_path -i $instrain_geneinfo_tsv_path -m $MicrobeCensus_summary_file -o $outdir -v -V
```

## Week 9 prep (5 minutes)

Next week, we will work on data visualization with ggplot2.
