# Introduction to Dimensionality Reduction

Week 5 goals are to understand the tools and steps involved in presenting and clustering high-dimensional data.

Hopefully, you have been able to find the PREMIX metagenome file in our OneDrive directory:

- PREMIX_mg_biom.json: Biom formatted output of species level Bracken results for all metagenome files
- PREMIX_metadata.csv: Metadata file for PREMIX metagenome files

## 1. phyloseq package documentation

- [phyloseq full stack](https://f1000research.com/articles/5-1492/v1)

## 2. Dimensionality reduction

The goal of dimensionality reduction is a way to summarize high dimensional data (i.e. many rows, many columns) in a way that is meaningful and statistically appropriate.

## 3. Bray-Curtis PCoA Analysis in Phyloseq

We will perform these steps in phyloseq since they make it fairly simple to do so. Since phyloseq is mostly a way to consistently structure the typical components of microbiome datasets and includes many additional R packages, you could (and probably should at some point!) do this with each component step but that is outside of our scope.

``` r
# load phyloseq
library(phyloseq)

########### import data
# import biom file to phyloseq
# see https://www.cell.com/cms/10.1016/j.isci.2020.100905/attachment/d3d5d291-4f11-4ecf-af92-558af5d1f474/mmc7
biom_path   <-"${path_to_biom_file}"
data        <- import_biom(paste0(biom_path,"PREMIX_mg_biom.json"), parseFunction=parse_taxonomy_default)
```

``` r
########### prepare data
# prepare sample data (i.e. metadata) for storage in phyloseq object
sampledata  <- sample_data(pm_metadata)

# pull OTU table from imported biom file
OTU         <- otu_table(data@otu_table, taxa_are_rows = TRUE)

# pull taxonomy table from imported biom file
TAX         <- tax_table(data@tax_table)

# use OTU table sample names to label sample data object
sample_names(sampledata)  <- sample_names(OTU)
```

``` r
########### make phyloseq object
physeq      <- phyloseq(TAX, sampledata, OTU)

    # label taxonomy table columns
    colnames(tax_table(physeq)) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
```

``` r
########### pre-process/clean phyloseq object
# remove taxa with unassigned sequences
physeq <- subset_taxa(physeq, Genus != "-1")

# remove taxa not seen more than 2 times
physeq <- filter_taxa(physeq, function(x) sum(x > 2) > (0.2*length(x)), TRUE)
  
# standardize abundances to median sequencing depth
    total <- median(sample_sums(physeq))
    standf <- function(x, t=total) round(t * (x / sum(x)))
    physeq_s <- transform_sample_counts(physeq, standf)
```

## Week 6 prep (5 minutes)

Next week, we will work on contig assembly and binning. To prepare for next week, let's:

- Read the [metaSPAdes paper](https://pubmed.ncbi.nlm.nih.gov/28298430/) 
