# Coverage calculations

Week 8 goals are to understand the tools and steps involved in calculating coverage statistics for genomes and genes with metagenomes of interest.

## 1. Calculating coverage statistics

Coverage describes how well represented a sequence (i.e. gene or genome) is within a sequencing file.

- [inStrain paper](https://doi.org/10.1038/s41587-020-00797-0)
- [MicrobeCensus paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0611-7)

## 2. Running inStrain

To run inStrain, you need a metagenome, reference [e.g. genome, set of genomes, or large set of genomes (i.e. a database)], and typically also a file containing genes predicted from your reference. If you start with the genome(s) and metagenome of interest, the [detailed inStrain documentation](https://instrain.readthedocs.io/en/latest/index.html) can help you run inStrain to calculate coverage statistics.


## 3. Running Microbe Census

Microbe Census is a tool that can estimate the average genome size and number of genome equivalents in your metagenome. This is useful because the size of the genomes in a metagenome could influence how well you have accurately sampled your population of interest.

``` console
######### define variables
indir=                        # directory containing paired reads from trimmomatic
outdir=                       # directory to write MicrobeCensus output, e.g. 06.microbecensus
ID=                           # pick one from the list of PM06 IDs when running an individual file
args="-n 100000000 -t 16"     # sample 100,000,000 million reads and run with 16 threads

######### at this point, you may have compressed your reads or changed your 
######### labeling if your read naming matches one of the conventions below, 
######### this will assign it to the R1 and R2 variable.

      # set reads by indir/id
        if [[ -f "${indir}/${ID}_1.fastq.gz" ]]; then
                
                R1=${indir}/${ID}_1.fastq.gz
                R2=${indir}/${ID}_2.fastq.gz

        elif [[ -f "${indir}/${ID}_P1.fastq.gz" ]]; then

                R1=${indir}/${ID}_P1.fastq.gz
                R2=${indir}/${ID}_P2.fastq.gz

        elif [[ -f "${indir}/${ID}_paired_1.fastq" ]]; then

                R1=${indir}/${ID}_paired_1.fastq
                R2=${indir}/${ID}_paired_2.fastq

        else
                R1=${indir}/${ID}_P1.fastq
                R2=${indir}/${ID}_P2.fastq

        fi

```

``` console
run_microbe_census.py $args ${R1},${R2} ${outdir}/${ID}.census
```


## Week 9 prep (5 minutes)

Next week, we will work on data visualization with ggplot2.
