# Metagenome quality control

Week 2 goals are to understand the tools and steps involved in quality control for metagenomic short reads.

---

## Trimmomatic Manuscript (10-20 Minutes)

[Trimmomatic](https://pubmed.ncbi.nlm.nih.gov/24695404/) is a tool that very widely used for preprocessing reads.

---

## Trim Those Reads (35-45 Minutes)

### Download PM06 Reads

[PM06S](https://emory-my.sharepoint.com/:u:/g/personal/mwoodwo_emory_edu/Efw2c9seTzVAgpeM8smB4HMBZI9REQ1spmZUbKlRvoVpew?e=RS43cp) is a file that contains contatenated raw reads.

### Sequencing by synthesis

[Illumina short-read high-throughput multiplexed sequencing](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwjXkJyQyvT8AhUWmIkEHYlsArsQtwJ6BAgKEAI&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DfCd6B5HRaZ8&usg=AOvVaw0qqyiPhh-rg-gJvFyHWMf8) is the most widely used sequencing approach.

This approach requires post-processing steps prior to analysis. When paired-end reads come off the flow cell as fastq data files from multiple flow cell lanes, they need to be concatenated as forward and reverse read files (R1/R2).

### Decompress Reads

``` console
tar -xf PM06S.tar.gz
```

### Run Trimmomatic

*Make sure you have installed conda, created a trimmomatic environment, and installed trimmomatic first!*

- if trimmomatic was installed with conda, you may need to specify the path of illumina adaptor sequences, which are also on github:

```console
# download trimmomatic repository
git clone https://github.com/timflutre/trimmomatic

# adapters are then found in trimmomatic/adapters
adapter_path=

# add this directory to your path so they are easy to locat (optional)
export PATH="$PATH:${adapter_path}"
```

- trimmomatic can then be run:

``` console
# set values for variables in generalized code
IDlist=/home/mwoodwo/prj/Berryhill/99.lists/IDlist.txt
indir=/home/mwoodwo/prj/Berryhill/00.raw
outdir=/home/mwoodwo/prj/Berryhill/01.trimmomatic

# with comma-delimited ID file SAMPLE,SRR_ACC
for line in `cat $IDlist`; do ID=$(echo $line | cut -d, -f1); echo starting $ID ...; R1=${indir}/${ID}/${ID}_1.fastq; R2=${indir}/${ID}/${ID}_2.fastq; trimmomatic PE -threads 10 -phred33 -summary ${outdir}/summary_files/${ID}_summary.txt $R1 $R2 ${outdir}/${ID}_P1.fastq ${outdir}/${ID}_U1.fastq ${outdir}/${ID}_P2.fastq ${outdir}/${ID}_U2.fastq ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36; echo $ID seems complete ...; echo ''; done
```

- unpaired reads can then be concatenated:

``` console
for line in `cat $IDlist`; do ID=$(echo $line | cut -d, -f1); echo starting $ID ...; cat ${outdir}/${ID}_U1.fastq ${outdir}/${ID}_U2.fastq > ${outdir}/${ID}_U.fastq; echo $ID appears complete ...; echo ''; done
```

---

## Other Quality Control Steps

- FastQC
- MetaQuast
- Decontamination of host reads with bowtie2 or BMTagger
- Kneaddata (wrapper for trimmomatic and human read removal)

---

## Week 3 prep (5 minutes)

Next week, we will work on short read taxonomic classification using kraken2 and bracken. A [(relatively) recent benchmarking study](https://www.sciencedirect.com/science/article/pii/S0092867419307755) supports the use of kraken2/bracken workflows for short read classification. To prepare for next week, let's:

- Read the [kraken2](https://pubmed.ncbi.nlm.nih.gov/31779668/) and [bracken](https://peerj.com/articles/cs-104/) papers
- create environments for kraken2 and bracken
- download the small version of the latest kraken2 database
- classify your trimmed metagenome reads from trimmomatic from week 2 from PM06
- [OPTIONAL] you may want to also look at [this recent paper from the kraken group](https://www.nature.com/articles/s41596-022-00738-y).
